src/Microsoft.OData.Core/Evaluation/ODataResourceMetadataContext.cs

```diff
@@ -169,60 +169,63 @@ public IODataResourceTypeContext TypeContext
         /// <param name="requiresId">Whether key properties are required to be returned</param>
         /// <returns>Key value pair array</returns>
-        internal static KeyValuePair<string, object>[] GetKeyProperties(
+        internal static IList<KeyValuePair<string, object>> GetKeyProperties(
             ODataResourceBase resource,
             ODataResourceSerializationInfo serializationInfo,
             IEdmEntityType actualEntityType,
             bool requiresId)
         {
             Debug.Assert(resource != null, "GetKeyProperties called for a null resource.");
 
-            KeyValuePair<string, object>[] keyProperties = null;
-            string actualEntityTypeName = string.IsNullOrEmpty(resource.TypeName) ? actualEntityType?.FullName() : resource.TypeName;
+            IList<KeyValuePair<string, object>> keyProperties = null;
 
             // if we have serializationInfo, try that first
             if (serializationInfo != null)
             {
-                keyProperties = ODataResourceMetadataContextWithoutModel.GetPropertiesBySerializationInfoPropertyKind(resource, ODataPropertyKind.Key, actualEntityTypeName);
+                keyProperties = ODataResourceMetadataContextWithoutModel.GetPropertiesBySerializationInfoPropertyKind(resource, ODataPropertyKind.Key, actualEntityType);
             }
 
             // if we didn't get any keys from serializationInfo, try using entity type
-            if ((keyProperties == null || keyProperties.Length == 0) && actualEntityType != null)
+            if ((keyProperties == null || keyProperties.Count == 0) && actualEntityType != null)
             {
-                keyProperties = GetPropertyValues(actualEntityType.Key(), resource, actualEntityType, requiresId).ToArray();
+                keyProperties = GetPropertyValues(actualEntityType.Key(), resource, actualEntityType, requiresId);
             }
 
-            if (!ValidateEntityTypeHasKeyProperties(keyProperties, actualEntityTypeName, requiresId))
+            if (!ValidateEntityTypeHasKeyProperties(keyProperties, resource, actualEntityType, requiresId))
             { 
                 return Enumerable.Empty<KeyValuePair<string, object>>().ToArray();
             }
 
             return keyProperties;
         }
 
-        private static IEnumerable<KeyValuePair<string, object>> GetPropertyValues(IEnumerable<IEdmStructuralProperty> properties, ODataResourceBase resource, IEdmEntityType actualEntityType, bool isRequired)
+        private static List<KeyValuePair<string, object>> GetPropertyValues(IEnumerable<IEdmStructuralProperty> properties, ODataResourceBase resource, IEdmStructuredType actualEntityType, bool isRequired)
         {
-            string actualEntityTypeName = actualEntityType.FullName();
             object primitiveValue;
+            // Initialize capacity to 1 because this method is mostly called to get values of key properties
+            // and most entities have a single key property

```

Comment:
What happens in cases where where entities have more than 1 property?

Reply:
Then the list will be resized. Here I was just optimizing for the common case.